modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/block/BlockService.scala

```diff
@@ -0,0 +1,222 @@
+          allBlocks <- blocks.toMap
+          allTips = allBlocks.collect { case (hash, ab @ AcceptedBlock(_, _, _, true)) => hash -> ab }
+          parentHashToTip = hashedBlock.signed.value.parent.map(ref => ref.hash -> allTips.get(ref.hash)).toList.toMap
+          tipParents = parentHashToTip.collect { case (hash, Some(accepted)) => hash -> accepted }
+          notTipParents = (parentHashToTip -- tipParents.keySet).keySet
+          tipParentsSize = tipParents.size
+          tipParentsToRemove = tipParents.filter {
+            case (_, block) =>
+              val enoughTips = allTips.size >= tipsConfig.minimumTipsCount + tipParentsSize
+              val tooManyTips = allTips.size >= tipsConfig.maximumTipsCount + tipParentsSize
+              val parentReachedUsageLimit = (block.usages + hashedBlock.proofsHash).size >= tipsConfig.maximumTipUsages
+              enoughTips && (tooManyTips || parentReachedUsageLimit)
+          }
+          tipParentsToRemain = tipParents -- tipParentsToRemove.keySet
+          remainingTips = allTips -- tipParentsToRemove.keySet
+          minTipHeight = remainingTips.values.map(_.block.signed.value.height.coerce).minOption.getOrElse(0L)
+          addAsTipCondition = remainingTips.size < tipsConfig.maximumTipsCount && minTipHeight < hashedBlock.signed.value.height.coerce
+          _ <- updateNoTipParents(notTipParents.toList, hashedBlock)
+          _ <- updateTipParentsToRemove(tipParentsToRemove.keySet.toList, hashedBlock)
+          _ <- updateTipParentsToRemain(tipParentsToRemain.keySet.toList, hashedBlock)
+          _ <- tryAddAsATip(addAsTipCondition, hashedBlock)
+        } yield ()
+
+      def pullTips(tipsCount: PosInt): F[Option[Tips]] =
+        for {
+          allBlocks <- blocks.toMap
+          allTips = allBlocks.collect { case (_, AcceptedBlock(block, _, _, true)) => block }.toList
+            .sortBy(_.signed.value.height.coerce)
+          pulledTips = allTips match {
+            case tips if tips.size >= tipsCount =>
+              NonEmptyList
+                .fromList(tips.take(tipsCount))
+                .map(parents => block.Tips(parents.map(p => BlockReference(p.proofsHash, p.signed.value.height))))
+            case _ => None
+          }
+        } yield pulledTips
+
+      private def isBlockAccepted(blockReference: BlockReference): F[Boolean] =
+        blocks(blockReference.hash).get.map(_.exists(_.isInstanceOf[AcceptedBlock]))
+
+      private def updateNoTipParents(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, false)) =>
+                (accepted.addUsage(hashedBlock.proofsHash).some, ().asRight)
+              case unexpected =>
+                (unexpected, NoTipAcceptedBlockNotFound(hash, unexpected).asLeft)
+            }
+            .flatMap(_.liftTo[F])
```

Comment:
what if it fails? one parent will have usage added, second not?

Reply:
If the first parent fails the update for the second one won't be attempted, that's right. Not 100% sure if that's how it should be.

Reply:
If any of the parents are invalid, the block is invalid. So we should only accept both or none

Reply:
And yes if the first one passes and the second fails we won't rollback the update for the first parent - so it's not atomic. Don't know if we could even make it atomic. At this point the block is validated and these cases shouldn't happen - if they happen it means we have some bug in our code and we need to fix it. I don't see a way to do it any better at this point. For the case of the tips usage it's not crucial, but in general it's the matter of atomic state shifts.

Reply:
If any of the parents are invalid, the block is invalid. So we should only accept both or none